### PR TITLE
TECH-477: Updated cucumber tests for /list_clients endpoint

### PR DESCRIPTION
**Pull request by KarolinaKopacz**
Need to update tests for /list_client endpoint.

TICKET: https://govstack-global.atlassian.net/browse/TECH-477

### CODE REVIEW

File: test/openAPI/features/list_clients.feature [hunk 1]
```diff
@@ -1,13 +1,29 @@
...
+    And The listClients endpoint response should match json schema
+
+  @unit @positive
   Scenario: Successfully retrieved the list of clients from GovStack with optional parameters in the request
-    Given Wants to retrieve the list of clients from GovStack with optional parameters specified
-    When The request with optional parameters to retrieve the list of Clients of GovStack is sent
-    Then The operation returns the list of Clients of GovStack
+    Given User wants to retrieve the the list of Clients of GovStack
+    When User sends GET request
+    And User provides optional query parameters "<serviceId>" as serviceId, "<instanceId>" as instanceId and optional header "<X-GovStack-Client>" as X-GovStack-Client
```
Comment (aleksander-reitsakas) [inline test/openAPI/features/list_clients.feature]: listClients have only one optional parameter <instanceId>.
`X-GovStack-Client` header is not needed. `Accept: application/json` header is needed.

File: test/openAPI/features/support/list_clients.js [hunk 2]
```diff
@@ -1,50 +1,67 @@
...
+Then('The listClients endpoint response should match json schema', async () => {
+  chai.expect(specListClients._response.json).to.be.jsonSchema(responseSchema);
+});
+
+// Scenario:Successfully retrieved the list of clients from GovStack with optional parameters in the request
 When(
-  'The request with optional parameters to retrieve the list of Clients of GovStack is sent',
-  () =>
+  'User provides optional query parameters {string} as serviceId, {string} as instanceId and optional header {string} as X-GovStack-Client',
+  (serviceId, instanceId, XGovStackClient) =>
```
Comment (aleksander-reitsakas) [inline test/openAPI/features/support/list_clients.js]: See previous comment

**Review by aleksander-reitsakas — CHANGES_REQUESTED** [review]

File: test/openAPI/features/list_clients.feature (hunk 1 above)
Comment (KarolinaKopacz) [reply test/openAPI/features/list_clients.feature]: Okay, thank you @aleksander-reitsakas ! I'm going to fix it

File: test/openAPI/features/list_clients.feature [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
-  Request endpoint: GET /listClients
+  @smoke
+  Scenario: Successfully retrieved the list of clients from GovStack
+
+    Given User wants to retrieve the the list of Clients of GovStack
+    When User sends GET request with given "serviceId" as serviceId, "instanceId" as instanceId and header
+    Then User receives a response from the listClients endpoint
+    And The listClients endpoint response should be returned in a timely manner
+    And The listClients endpoint response should have status 200
+    And The listClients endpoint response should have content-type header
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/list_clients.feature]: ```The listClients endpoint response should have content-type header``` sounds a bit confusing i think we should specify the header that we check

File: test/openAPI/features/list_clients.feature [hunk 4]
```diff
@@ -1,13 +1,30 @@
...
-  Scenario: Successfully retrieved the list of clients from GovStack with optional parameters in the request
-    Given Wants to retrieve the list of clients from GovStack with optional parameters specified
-    When The request with optional parameters to retrieve the list of Clients of GovStack is sent
-    Then The operation returns the list of Clients of GovStack
+    Given User wants to retrieve the the list of Clients of GovStack
+    When User sends GET request with given "<serviceId>" as serviceId, "<instanceId>" as instanceId and header
+    Then User receives a response from the listClients endpoint
+    And The listClients endpoint response should be returned in a timely manner
+    And The listClients endpoint response should have status 200
+    And The listClients endpoint response should have content-type header
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/list_clients.feature]: ```The listClients endpoint response should have content-type header``` sounds a bit confusing i think we should specify the header that we check

File: test/openAPI/features/list_clients.feature [hunk 5]
```diff
@@ -1,13 +1,30 @@
+@method=GET @endpoint=/listClients
 Feature: This endpoint is used to retrieve the list of clients from GovStack.
 
-  Request endpoint: GET /listClients
+  @smoke
+  Scenario: Successfully retrieved the list of clients from GovStack
+
+    Given User wants to retrieve the the list of Clients of GovStack
+    When User sends GET request with given "serviceId" as serviceId, "instanceId" as instanceId and header
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/list_clients.feature]:
```suggestion
    When User sends GET request with given "serviceId" as serviceId, "instanceId" as instanceId
```

File: test/openAPI/features/list_clients.feature [hunk 6]
```diff
@@ -1,13 +1,30 @@
...
-    Given Wants to retrieve the the list of Clients of GovStack
-    When The request to retrieve the list of Clients of GovStack is sent
-    Then The operation returns the list of Clients of GovStack
-
-  Scenario: Successfully retrieved the list of clients from GovStack with optional parameters in the request
-    Given Wants to retrieve the list of clients from GovStack with optional parameters specified
-    When The request with optional parameters to retrieve the list of Clients of GovStack is sent
-    Then The operation returns the list of Clients of GovStack
+    Given User wants to retrieve the the list of Clients of GovStack
+    When User sends GET request with given "<serviceId>" as serviceId, "<instanceId>" as instanceId and header
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/list_clients.feature]:
```suggestion
    When User sends GET request with given "<serviceId>" as serviceId, "<instanceId>" as instanceId
```

File: test/openAPI/features/support/helpers/helpers.js [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
   localhost: 'http://localhost:3366/',
+  listClientsEndpoint: 'listClients',
   getOpenApiEndpoint:
     '{GovStackInstance}/{memberClass}/{memberCode}/{applicationCode}/getOpenAPI',
   defaultExpectedResponseTime: 15000,
   getOpenApiExpectedSchema: {
     type: 'string',
   },
+  acceptHeader: {
+    key: 'Accept',
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/helpers/helpers.js]:
```suggestion
    key: 'content-type',
```

File: test/openAPI/features/support/list_clients.js [hunk 8]
```diff
@@ -1,50 +1,88 @@
...
 
-When('The request to retrieve the list of Clients of GovStack is sent', () =>
-  specListClients.get(baseUrl)
+When(
+  'User sends GET request with given {string} as serviceId, {string} as instanceId and header',
+  (serviceId, instanceId) =>
+    specListClients
+      .get(baseUrl)
+      .withHeaders(acceptHeader.key, acceptHeader.value)
+      .withQueryParams({
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]: I'd remove .withHeaders and check just response if there is valid content-type

File: test/openAPI/features/list_clients.feature [hunk 9]
```diff
@@ -1,13 +1,30 @@
...
+    Given User wants to retrieve the the list of Clients of GovStack
+    When User sends GET request with given "serviceId" as serviceId, "instanceId" as instanceId and header
+    Then User receives a response from the listClients endpoint
+    And The listClients endpoint response should be returned in a timely manner
+    And The listClients endpoint response should have status 200
+    And The listClients endpoint response should have content-type header
+    And The listClients endpoint response should match json schema
 
+  @unit @positive
   Scenario: Successfully retrieved the list of clients from GovStack
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/list_clients.feature]:
```suggestion
  Scenario Outline: Successfully retrieved the list of clients from GovStack
```

File: examples/mock/mockoon-bb-information-mediator.json [hunk 10]
```diff
@@ -24,15 +24,38 @@
           "filePath": "",
           "databucketID": "",
           "sendFileAsBody": false,
-          "rules": [],
-          "rulesOperator": "OR",
+          "rules": [
+            {
+              "target": "header",
```
Comment (mgrochalskisoldevelo) [inline examples/mock/mockoon-bb-information-mediator.json]: i think this rule is not necessary

File: test/openAPI/features/list_clients.feature [hunk 11]
```diff
@@ -1,13 +1,30 @@
+@method=GET @endpoint=/listClients
 Feature: This endpoint is used to retrieve the list of clients from GovStack.
 
-  Request endpoint: GET /listClients
+  @smoke
+  Scenario: Successfully retrieved the list of clients from GovStack
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/list_clients.feature]: Should scenarios have the same title? 

```suggestion
  Scenario: Successfully retrieved the list of clients from GovStack smoke type test
```

File: test/openAPI/features/support/list_clients.js [hunk 12]
```diff
@@ -1,50 +1,88 @@
...
+      .should.have.header('content-type', acceptHeader.value);
+  }
+);
+
+Then('The listClients endpoint response should match json schema', async () => {
+  chai.expect(specListClients._response.json).to.be.jsonSchema(responseSchema);
+});
+
+// Scenario: Successfully retrieved the list of clients from GovStack
 When(
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]: This When is not necessary, both scenarios should call the first one

File: test/openAPI/features/support/list_clients.js [hunk 13]
```diff
@@ -1,50 +1,88 @@
...
+Then('The listClients endpoint response should have status 200', () => {
+  specListClients.response().to.have.status(200);
+});
+
+Then(
+  'The listClients endpoint response should have content-type header',
+  () => {
+    specListClients
+      .response()
+      .should.have.header('content-type', acceptHeader.value);
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]:
```suggestion
      .should.have.header(acceptHeader.key, acceptHeader.value);
```

after change in helpers file

File: test/openAPI/features/support/list_clients.js [hunk 14]
```diff
@@ -1,50 +1,88 @@
...
+  }
+);
+
+Then('The listClients endpoint response should match json schema', async () => {
+  chai.expect(specListClients._response.json).to.be.jsonSchema(responseSchema);
+});
+
+// Scenario: Successfully retrieved the list of clients from GovStack
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]:
```suggestion
// Scenario Outline: Successfully retrieved the list of clients from GovStack
```

**Review by mgrochalskisoldevelo — CHANGES_REQUESTED** [review]

File: test/openAPI/features/list_clients.feature [hunk 15]
```diff
@@ -1,13 +1,30 @@
...
+    And The listClients endpoint response should have content-type: application/json header
+    And The listClients endpoint response should match json schema
 
-  Scenario: Successfully retrieved the list of clients from GovStack with optional parameters in the request
-    Given Wants to retrieve the list of clients from GovStack with optional parameters specified
-    When The request with optional parameters to retrieve the list of Clients of GovStack is sent
-    Then The operation returns the list of Clients of GovStack
+  @unit @positive
+  Scenario Outline: Successfully retrieved the list of clients from GovStack
+    Given User wants to retrieve the the list of Clients of GovStack
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/list_clients.feature:18]: add empty line above Given for consistency

**Review by mgrochalskisoldevelo — CHANGES_REQUESTED** [review]

**Review by mgrochalskisoldevelo — APPROVED** [review]

File: test/openAPI/features/support/list_clients.js [hunk 16]
```diff
@@ -1,50 +1,74 @@
...
+  (serviceId, instanceId) =>
+    specListClients.get(baseUrl).withQueryParams({
+      serviceId: serviceId,
+      instanceId: instanceId,
+    })
 );
 
-Then('The operation returns the list of Clients of GovStack', async () => {
-  specListClients.expectStatus(200).expectJsonSchema(responseSchema);
+Then('User receives a response from the listClients endpoint', async () => {
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]: unnecessary curly brackets

File: test/openAPI/features/support/list_clients.js [hunk 17]
```diff
@@ -1,50 +1,74 @@
...
   await specListClients.toss();
 });
 
-// Scenario: Successfully retrieved the list of clients from GovStack with optional parameters in the request
-Given(
-  'Wants to retrieve the list of clients from GovStack with optional parameters specified',
-  () => 'Optional parameters for the route were specified'
+Then(
+  'The listClients endpoint response should be returned in a timely manner',
+  () => {
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]: unnecessary curly brackets

File: test/openAPI/features/support/list_clients.js [hunk 18]
```diff
@@ -1,50 +1,74 @@
...
+    specListClients
+      .response()
+      .to.have.responseTimeLessThan(defaultExpectedResponseTime);
+  }
 );
 
-When(
-  'The request with optional parameters to retrieve the list of Clients of GovStack is sent',
-  () =>
+Then('The listClients endpoint response should have status 200', () => {
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]: unnecessary curly brackets

File: test/openAPI/features/support/list_clients.js [hunk 19]
```diff
@@ -1,50 +1,74 @@
...
+  specListClients.response().to.have.status(200);
+});
+
+Then(
+  'The listClients endpoint response should have content-type: application\\/json header',
+  () => {
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]: unnecessary curly brackets

File: test/openAPI/features/support/list_clients.js [hunk 20]
```diff
@@ -1,50 +1,74 @@
...
-        serviceId: 'string',
-        instanceId: 'string',
-      })
+      .response()
+      .should.have.header(acceptHeader.key, acceptHeader.value);
+  }
 );
 
-// "Then" already written in line 83-86
+Then('The listClients endpoint response should match json schema', async () => {
```
Comment (mgrochalskisoldevelo) [inline test/openAPI/features/support/list_clients.js]: unnecessary curly brackets + is async needed in this function?

**Review by mgrochalskisoldevelo — CHANGES_REQUESTED** [review]
During ```listMethods``` endpoint review I found a few small things that should be changed for consistency.

**Review by aleksander-reitsakas — APPROVED** [review]

**Review by mgrochalskisoldevelo — APPROVED** [review]

**Pull request merged**